### ISSUE TITLE
Community Presets implementation

### ISSUE THREAD

**Issue by Digitalone1** [wwmm/easyeffects#2445]
After reading #1771, I wanted to test presets placed in a system config folder.

In addition to the common local config `~/.config/easyeffects`, EE looks into system folders given by `g_get_system_config_dirs` (appending input/output directories): https://github.com/wwmm/easyeffects/blob/master/src/presets_manager.cpp#L34-L38

And also into `etc/easyeffects/input` and `etc/easyeffects/output`: https://github.com/wwmm/easyeffects/blob/master/src/presets_manager.cpp#L41-L42

* The first issue I encountered is that if I copy a preset into a system folder, the new preset name is not shown into the menu. But if I shut down and relaunch the application, the new presets are listed inside the menu. It seems system folders are scanned only at the application startup. @wwmm can you confirm?

* Another issue is that the menu shows the delete button for a system preset, but this does not work because the application does not have root privileges to erase the related file.

System folders could be useful in the future if someone wants to distribute community presets in a separate package. But the issues above should be addressed.

- If the presets are installed when EE is running, they should be immediately visible, not requiring to restart the application.
- System presets should net be removable from the preset menu because they are tied to the external package, so the should disappear only if the user uninstalls the package. Besides the remove button does not work for them and this is confusing for the user. Therefore we should remove the delete button for system preset entries.
- We should test what happens for double presets (files with the same name) placed in both user and system folders. I suppose the ones in user config should take the precedence.

**Comment by wwmm** [comment]
> It seems system folders are scanned only at the application startup. @wwmm can you confirm?

Yes. At this moment we listen to changes only in the user folder and not in the system https://github.com/wwmm/easyeffects/blob/cf37bb3b9c692850b93f299eaeeaff874f4f75ab/src/presets_manager.cpp#L65C13-L65C13. THE problem is that while the user config is in only one place the system preset path can have multiple locations. I am not sure if creating a folder monitor for each possible path is a good approach or even a good idea from a performance point of view.

>System presets should net be removable from the preset menu because they are tied to the external package, so the should disappear only if the user uninstalls the package. Besides the remove button does not work for them and this is confusing for the user. Therefore we should remove the delete button for system preset entries.

That is why I think that community or system presets should be tied to the `import dialog`. This way the workflow forces the user to import everything to a single directory where we can actually delete files. The import dialog could point to a system path by default. And in this system path the community presets would be located for people interested in them.

>We should test what happens for double presets (files with the same name) placed in both user and system folders. I suppose the ones in user config should take the precedence.

I am not sure about what would happen. We remove duplicated names from the list. So it will depend on the order `erase` is removing the duplicated values.

**Comment by Digitalone1** [comment]
> That is why I think that community or system presets should be tied to the import dialog. This way the workflow forces the user to import everything to a single directory where we can actually delete files. The import dialog could point to a system path by default. And in this system path the community presets would be located for people interested in them.

Can you explain this method more in depth? I don't get it fully.

Anyway, aside of the community presets, if we still want to support system folders, we still need to hide the remove button because the file cannot be deleted.

If this is troublesome to do, then we should remove system folders. 

About the scanning, we could just support `\etc\easyeffects` since the others seem to be unused anyway. I see lots of softwares installing system configuration files a custom folder under `etc`, not specific subfolder or other locations.

**Comment by wwmm** [comment]
>Can you explain this method more in depth? I don't get it fully

It isn't really a method because all we would be doing would be to set the default path of the import dialog window to the system presets folder and nothing else. At this moment we do not set any default path to that dialog. Let's imagine that package maintainers start to put community presets inside `/etc/easyeffects/presets`. If the import dialog window starts there by default people will see what is there and choose to import one if they want to do so.

>Anyway, aside of the community presets, if we still want to support system folders, we still need to hide the remove button because the file cannot be deleted.

Sure. But I am not sure if these system folders are being used by many people. Even after all these years I can't remember cases where people are actually using presets in system folders.

>About the scanning, we could just support \etc\easyeffects since the others seem to be unused anyway. I see lots of softwares installing system configuration files in a custom folder under etc, not specific subfolder or other locations.

I think that the difficulty here is that for presets subfolders make sense. And the `new / changed / removed` monitoring probably does not work recursively. So the monitoring would still be "broken".

**Comment by Digitalone1** [comment]
It's a simple idea and it might work. The only issue is that if I usually download presets in a local folder (i.e. the `Download` directory), then having the import dialog to start always from the system folder it's a bit of an hassle.

If I remember correctly, it starts from the folder where it has been left the last time it was used, am I correct?

Anyway, if we go on that route, system folders should be removed from the code and `/etc/easyeffects` would be the folder where packages should install community presets (which I think on Flatpak it would be something under `/var/`.

Besides we should also allows multiple selections, because if a package installs 20 presets and I'd like to test them all, importing one at a time would be a big pain. Honestly, I don't remember if we already allow multiple selection.

**Comment by wwmm** [comment]
>It's a simple idea and it might work. The only issue is that if I usually download presets in a local folder (i.e. the Download directory), then having the import dialog to start always from the system folder it's a bit of an hassle.

Yes. It is annoying. But at least for me it is already annoying because it does not start in the folder I usually copy presets before importing.

>It's a simple idea and it might work. The only issue is that if I usually download presets in a local folder (i.e. the Download directory), then having the import dialog to start always from the system folder it's a bit of an hassle.

At least on GNOME it shows the `Recent files` list by default and not a directory.

>Anyway, if we go on that route, system folders should be removed from the code and /etc/easyeffects would be the folder where packages should install community presets (which I think on Flatpak it would be something under /var/.

Yes. But I suggest `/etc/easyeffects/presets` because maybe in the future the community decides to do something similar for impulse response files. 

>Besides we should also allow multiple selections, because if a package installs 20 presets and I'd like to test them all, importing one at a time would be a big pain. Honestly, I don't remember if we already allow multiple selection.

This should be possible. If I remember well the file dialog has a multiple files selection mode.

**Comment by wwmm** [comment]
I will take a look at the documentation. But thinking about the path again we probably should use `/usr/share/easyeffects/presets`. The `/etc` path is historically more for system services configuration. For example gnome apps put this kind of files under `/usr/share`.

**Comment by wwmm** [comment]
>Creating those directories after the installation (otherwise we can't start from them in import dialogs if they don't exist; @wwmm should we implement a post install script for this?)

I have updated the master branch using a Meson function designed for this task https://github.com/wwmm/easyeffects/commit/ab8c4fe1269468faa296a43de0d31bf767e3270f

**Comment by wwmm** [comment]
Inside the build time generated header `#include "config.h"` there will be `#define SYSTEM_PRESETS_DIR`

**Comment by wwmm** [comment]
> Adding multiple selection support for import dialogs related to preset and impulse managers (@wwmm any hint on that?);

We just have to see how to use [gtk_file_dialog_open_multiple](https://docs.gtk.org/gtk4/method.FileDialog.open_multiple.html) instead of the `gtk_file_dialog_open` we use right now.

**Comment by wwmm** [comment]
>Let the import dialogs start from the proper directories where community presets should be placed (@wwmm any hint here?);

It seems that we just have to set https://docs.gtk.org/gtk4/property.FileDialog.initial-folder.html. But I did not test it.

**Comment by vchernin** [comment]
To be clear the idea here is these preset packages would be installed with e.g. pacman and then the presets be shown automatically in easyeffects' dialog right? I think this can work well in flatpak but I will need to add some new configuration to the flatpak package.

In flatpak we will need use a different directory. Since it is configurable that isn’t an issue. Flatpak packages normally cannot modify things `/usr` in the flatpak package or on the host system.

The best reference I have is currently the flatpak build finds plugins at `/app/extensions/Plugins/lv2`. We would probably want something like `/app/extensions/Presets`. Unfortunately there isn’t much documentation for how this extension system works but from examples it is easy enough to understand.

https://github.com/wwmm/easyeffects/blob/ab8c4fe1269468faa296a43de0d31bf767e3270f/util/flatpak/com.github.wwmm.easyeffects.Devel.json#L13

https://github.com/wwmm/easyeffects/blob/ab8c4fe1269468faa296a43de0d31bf767e3270f/util/flatpak/com.github.wwmm.easyeffects.Devel.json#L36

On a related note it would probably be good to have at least 1 reference preset package by the time this feature is released? And to also test it.

**Comment by wwmm** [comment]
>Implementing a guideline for packagers to be published here on Github explaining how to provide community presets (where to place them, how to do for avoiding conflicts, etc...);

You are right. Now that I think about it inside `easyeffects/presets` we will have to create a folder for `input` and another for `output` presets.

>On a related note it would probably be good to have at least 1 reference preset package by the time this feature is released? 

It is fine. But I do not have presets to put there. The ones I have are very specific to my hardware and use cases. Maybe the ones @Digitalone1 already shares in her github page could be installed by default as examples. If I am not mistaken @p-chan5 is also maintaining a github repository with presets. Maybe he is interested in this.

Github has a feature I never used that allows third party git projects to be fetched as subprojects. Maybe if we look at it carefully there is a way to make Meson directly get the presets from @Digitalone1 and @p-chan5 at build time without the need to insert all of this in EasyEffects folder.

>In flatpak we will need use a different directory. Since it is configurable that isn’t an issue. Flatpak packages normally cannot modify things /usr in the flatpak package or on the host system.

Shouldn't https://github.com/wwmm/easyeffects/commit/ab8c4fe1269468faa296a43de0d31bf767e3270f already be putting things in a place visible to Flatpak?

**Comment by vchernin** [comment]
> Git has a feature I never used that allows third party git projects to be fetched as subprojects. Maybe if we look at it carefully there is a way to make Meson directly get the presets

Wouldn’t it be better to not have any presets at build time? And instead make some preset packages recommended pacman dependencies of easyeffects? So users can uninstall them easily, and more importantly it doesn’t affect the build at all?

> Shouldn't https://github.com/wwmm/easyeffects/commit/ab8c4fe1269468faa296a43de0d31bf767e3270f already be putting things in a place visible to Flatpak?

Well it is visible to the easyeffects flatpak package, but not to other flatpak packages or future flatpak extension packages. While system packages installed from e.g. pacman could technically modify the flatpak data directory, it is unusual for e.g. pacman to manage files for a flatpak package.

**Comment by wwmm** [comment]
>Wouldn’t it be better to not have any presets at build time? And instead make some preset packages recommended pacman dependencies of easyeffects? So users can uninstall them easily, and more importantly it doesn’t affect the build at all?

It is fine to do this. I thought you were thinking about shipping an example inside EasyEffects package. But your idea is to have thrid party distribution packages available by the time we release this EE feature. Ok.

>Well it is visible to the easyeffects flatpak package, but not to other flatpak packages or future flatpak extension packages. While system packages installed from e.g. pacman could technically modify the flatpak data directory, it is unusual for e.g. pacman to manage files for a flatpak package.

Oh... Ok.

**Comment by vchernin** [comment]
Another question, but should we establish a convention/recommendation on how to group/organize preset packages? Or is this best left to packagers to decide? What if most packages have 10 presets, but one packager decides to split 10 presets into individual packages? 

An example from https://github.com/wwmm/easyeffects/issues/1771#issuecomment-1235647058

> Should we have:
> 
> - A package containing every preset from the wiki:
> Arch: `easyeffects-all-presets` 
> Flatpak: `com.github.wwmm.easyeffects.Presets.all`
> 
> - One package per presets repo listed in the wiki:
> Arch: `easyeffects-presets-digitalone1`
> Flatpak: `com.github.wwmm.easyeffects.Presets.digitalone1`.
> 
> - One package per preset file listed in the wiki:
> Arch: `easyeffects-presets-digitalone1-loudnessequalizer`
> Flatpak: `com.github.wwmm.easyeffects.Presets.digitalone1-loudnessequalizer`.

From my earlier comment:

> On a related note it would probably be good to have at least 1 reference preset package by the time this feature is released?

If anyone would like to make a preset package early as a reference/testing package, I can contribute a flatpak manifest file and submit it to the flathub github when everything is ready.

**Comment by wwmm** [comment]
>Should we establish a convention/recommendation on how to group/organize preset packages? 

I would let this decision to the package maintainers. As in the end the user will have to use EE import dialog it does not really matter to us how the files will be inside the system `easyeffects/presets` directory.

**Comment by wwmm** [comment]
>it does not really matter to us how the files will be inside the system easyeffects/presets directory.

It is probably a good idea to separate input from output presets to make the user life easier. But other than this it should not matter much what is done.

**Comment by Digitalone1** [comment]
> Well it is visible to the easyeffects flatpak package, but not to other flatpak packages or future flatpak extension packages.

So, if I understood correctly, a flatpak package can install the presets in the proper place where EE flatpak can see them, right?

My proposal for the location. Whether to use `/usr` or `/etc`, `easyeffects` folder will contain:
- `input` for input effects
- `output` for output effects
- `irs` for convolver impulses
- `rnnoise` for RNNoise models

Any of the above folders can have custom user files. This is useful for users that may have more user profiles and they want to import presets from a session to another. So they will move the files in the system directory (using root) and will import them in the application running on the other user session. I.e `easyeffects/output/my-default-preset.json`.

The same folders can have custom directories that will belong to community packages. So If I have a package shipping presets and impulses I will install:
-  `easyeffects/output/community-package-name/community-out-preset.json`
-  `easyeffects/input/community-package-name/community-in-preset.json`
- `easyeffects/irs/community-package-name/community-conv-impulse.json`

This way there will be no conflict because every package should have its own subfolder. 

A package can install additional subfolders, i.e.:
- `easyeffects/irs/community-package-name/specific-device1/device1-impulse.json`
- `easyeffects/irs/community-package-name/specific-device2/device2-impulse.json` 

What do you think about this scheme?

Anyway, Easyeffects package will install the `easyeffects` directory, but on the removal the directory should be left because it may still contain community presets or user custom presets.

**Comment by Digitalone1** [comment]
> Wouldn’t it be better to not have any presets at build time? And instead make some preset packages recommended pacman dependencies of easyeffects? So users can uninstall them easily, and more importantly it doesn’t affect the build at all?

It's doable, but not all distro will follow. Maybe doing it for Flatpak should be a good idea since its usage will grow on time.

> If anyone would like to make a preset package early as a reference/testing package, I can contribute a flatpak manifest file and submit it to the flathub github when everything is ready.

Sure. Just pick some from [here](https://github.com/wwmm/easyeffects/wiki/Community-presets). I'd say @p-chan5, @JackHack96 and mine which seems more generic.

**Comment by wwmm** [comment]
>What do you think about this scheme?

It seems good.

**Comment by vchernin** [comment]
> So, if I understood correctly, a flatpak package can install the presets in the proper place where EE flatpak can see them, right?

Yes

> It's doable, but not all distro will follow. Maybe doing it for Flatpak should be a good idea since its usage will grow on time.

Including any preset packages at build time in my mind is too inconsistent. If we want to encourage preset authors/packagers to make their preset packages easily available they should make a distribution package right? But if we have several preset packages added at build time they will likely ask to add their preset to the easyeffects build. So then we need to decide what goes in at build time and what should go in a separate package. Instead, if we don’t include any presets at build time, it means there would be reference preset packages on distributions making it easier for others to follow. 

Not to mention build time presets may introduce an annoyance with the flatpak package. I am not sure if preset extensions installed from separate flatpak packages at runtime, and presets installed at build time could be easily combined into one directory. It is something I need to test but if easyeffects never includes presets at build time it avoids this problem.

> Any of the above folders can have custom user files. This is useful for users that may have more user profiles and they want to import presets from a session to another. So they will move the files in the system directory (using root) and will import them in the application running on the other user session.

Something worth noting is mixing package installed presets and user installed presets in a `/usr` directory could cause confusion. Usually `/usr` should only be modified by the system package manager and so should only contain package installed presets (ignoring `/usr/local`). On immutable distros like fedora silverblue, `/usr` is mounted as read only. You can only write to `/etc` or your home directory in `/var`.

With these changes can’t users add their own presets to the easyeffects config directory as done today? If they really wanted they can modify `/usr`, but most users probably shouldn’t. At least it could use a warning.

The rest of the proposal I quite like and I think it should work well.

**Comment by Digitalone1** [comment]
@vchernin I agree not adding presets at build time.

You said

> And instead make some preset packages recommended pacman dependencies of easyeffects? So users can uninstall them easily, and more importantly it doesn’t affect the build at all?

On this I don't know. Arch packagers should be contacted, but not all distro will follow. If it's doable on Flatpak, it's not an issue.

> Something worth noting is mixing package installed presets and user installed presets in a /usr directory could cause confusion.

Ok, so we would only recommend to install directories in `easyeffects` folder.

> On immutable distros like fedora silverblue, /usr is mounted as read only. You can only write to /etc or your home directory in /var.

I didn't know that. So we can't use `/usr` for community presets anyway.

> With these changes can’t users add their own presets to the easyeffects config directory as done today? If they really wanted they can modify /usr, but most users probably shouldn’t. At least it could use a warning.

Sure. I was only suggesting that users could move their own presets to `easyeffects/output` to make them available on another user session (because files of a specific user can't be accessed from another user session). But it's not mandatory, we could just ignore this use case.

**Comment by vchernin** [comment]
> Ok, so we would only recommend to install directories in easyeffects folder.

I am not sure what you mean by this?

> I didn't know that. So we can't use /usr for community presets anyway.

I think to some degree using `/usr/share/easyeffects` for package manager provided presets would be ideal, as that is what that directory is there for. But `/usr` is read only sometimes. I don't really see a better way besides only using `/etc/easyeffects` as the `easyeffects` directory (ignoring flatpak for a moment). It works for package managers and for manually copying files not available in a package, and means we can use the simplification of having the import dialog always point to it. 

Is this picture correct or am I missing something? Do you see pacman installed packages modifying  `~/.config/easyeffects/`?

One issue I am not realizing I did not think through.. the flatpak extensions directory, where the plugins will be intalled cannot actually be modified by the user. From within the easyeffects flatpak package the directory will be: `/app/extensions/Presets` and so you will have e.g. `/app/extensions/Presets/output/community-package-name/community-out-preset.json`. However, this directory is always created at runtime using the flatpak extension packages that are installed. Meaning, users cannot permanently modify this directory without creating a whole flatpak package. Is this a problem?  

I think it is a problem we can skip, as flatpaks would be the easiest to publish many packages for (it could be easily scripted). If we need to revise it in the future (and somehow allow multiple directories to host "community presets") it could be done without breaking backwards compatbility, since we will still need to use `/app/extensions/Presets` regardless.

**Comment by wwmm** [comment]
>I think to some degree using /usr/share/easyeffects for package manager provided presets would be ideal, as that is what that directory is there for. But /usr is read only sometimes. I don't really see a better way besides only using /etc/easyeffects/presets as the directory the import dialog points to (ignoring flatpak for a moment). It works for package managers and for manually copying files not available in a package.

EasyEffects will still use `~/.config/easyeffects/` and its equivalent in the Flatpak side. The only difference is the import dialog pointing by default to the system presets folder. I understand some users may want to put things in the system folder. But I do not see a read only `/usr` as a problem worth of a solution from our side. Traditionally `/etc` is more for things like services settings so I think it is better to not put things like presets there.

**Comment by vchernin** [comment]
> EasyEffects will still use `~/.config/easyeffects/` and its equivalent in the Flatpak side. The only difference is the import dialog pointing by default to the system presets folder. I understand some users may want to put things in the system folder. But I do not see a read only `/usr` as a problem worth of a solution from our side. Traditionally `/etc` is more for things like services settings so I think it is better to not put things like presets there.

I am fine either way, I don't think having  `~/.config/easyeffects/` still available and necessary for some situations should be too confusing.

So just to decide so we don't need to revisit this, should it be `/usr/share/easyeffects/presets` then? In flatpak it can be `/app/extensions/Presets`, which is similar to where many audio plugins sit in `/app/extensions/Plugins`.

When I have time I will prototype the flatpak setup and make sure there isn't any unexpected weirdness.

**Comment by wwmm** [comment]
>So just to decide so we don't need to revisit this, should it be /usr/share/easyeffects/presets then? In flatpak it can be /app/extensions/Presets, which is similar to where many audio plugins sit in /app/extensions/Plugins.

Yes. Inside this folder package maintainers will create subfolders to organize the presets as commented before by @Digitalone1.

**Comment by Digitalone1** [comment]
> I am not sure what you mean by this?

That we won't recommend users to install files in `/usr/share/easyeffects/output`, so it should contain only directories related to packages (even if a user can still move files there anyway).

**Comment by p-chan5** [comment]
Hi developers. Thank you for considering me in this idea. Sure, I'd like to help with this!

It would be nice if installing these packages were two or three clicks away, this would benefit both new and novice users. Of course, the results will depend on the hardware they have, although taking into account that the Linux audio drivers are "generic" the quality of the speaker, headphones or monitors will be what makes the difference.

I agree with this scheme. Perhaps I would be a little more careful with distributing `.iris` or `.rnnoise` files as they can be copyright infringing or in some cases confusing to newcomers.

Regarding the maintenance of this package, I could provide updates that are in sync with my [latest release](https://github.com/p-chan5/EasyPulse/releases/latest) and are compatible with the latest EE stable version.

**Comment by Digitalone1** [comment]
Most of the work is done. But, as I said in #2485, even if the initial folder is set, the GTK file dialog starts from home folder, so it's not working as intended.

**Comment by wwmm** [comment]
>even if the initial folder is set, the GTK file dialog starts from home folder, so it's not working as intended.

Unfortunately the same thing is happening on my computer :-(. Probably a GTK bug =/

**Comment by Digitalone1** [comment]
> Unfortunately the same thing is happening on my computer :-(. Probably a GTK bug =/

Should we fill an issue ticket on Gitlab?

**Comment by wwmm** [comment]
>Should we fill an issue ticket on Gitlab?

I am not sure. I have just found some interesting comments at https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/5393#note_1633736

**Comment by wwmm** [comment]
Even if change the dconf setting that the gtk developer was talking about the `initial-folder` property does not do what we want. It just goes to the user home folder instead =/. This parameter is useless as it is now...

**Comment by Digitalone1** [comment]
After a search on Google, I found this: https://gitlab.gnome.org/GNOME/xdg-desktop-portal-gnome/-/issues/84

Well, there's nothing to do. Just leave the things as they are and let's hope it will be fixed in the future...

**Comment by Digitalone1** [comment]
> I think we will have to think about another solution to point the user to the community presets.

Telling where to find them. But it's not very user friendly. 

Or we find a way to import the presets directly from EasyEffects, but don't know how. Maybe scanning the folder and showing in another widget the installed presets, the nyou can import what you want with a click. 

I wonder also if there a way to import a preset just dragging one or more files into the window (just like mpv which can open files by dragging). This way we can open the file explorer directly to the folder and the user will drag what they want. 

But I don't even know if it's possible to open the file explorer to a predefined position from a GTK app. From Atom/Pulsar editor it's possible, but it's an Electron app.

**Comment by wwmm** [comment]
>Maybe scanning the folder and showing in another widget the community presets installed, then you can import locally what you want with a click.

It is a way. The downside is having to do recursive scanning on our side to deal with subfolders. And as I doubt it will be possible to put all this in the presets menu this functionality may be located in a non obvious place of our window.

>I wonder also if there a way to import a preset just dragging one or more files into the window (just like mpv which can open files by dragging). This way we can open the file explorer directly to the folder and the user will drag what they want.

Probably possible. We already do drag and drop to change the effects order in the pipeline. 

>But I don't even know if it's possible to open the file explorer to a predefined position from a GTK app.

Good question. If there is the procedure is probably the same we used for the help https://github.com/wwmm/easyeffects/blob/8fd01a8ca3308b4568a78d47ed58c6672dc21aef/src/application.cpp#L439 bu with different arguments.

**Comment by Digitalone1** [comment]
@wwmm OK, but if we would go with the drag and drop method, it's very difficult to find a place where to drop the presets. There's the preset menu, but it's closed when a click is made outside the window. Conversely it would be easier for the RNNoise and the Convolver since the drop could be made inside the plugin box.

**Comment by wwmm** [comment]
>it's very difficult to find a place where to drop the presets.

Yes. I think that we would have to make the whole window a drop target. If that is viable as our code is right now or if it will cause side effects I do not know.

**Comment by Digitalone1** [comment]
> Yes. I think that we would have to make the whole window a drop target. If that is viable as our code is right now or if it will cause side effects I do not know.

Or maybe a special plugin box where we make a child box with the label `Drop here the community presets`. And another label with a link to open the folder (I hope it works also outside Gnome).

**Comment by servimo** [comment]
Maybe a button with both options: add preset or drag the file here.

**Comment by Digitalone1** [comment]
Reading the issues linked above, it seems that `initial-folder` property works only for the saving dialog, not for the opening one.

I don't think this will be solved shortly, and unfortunately it's a show stopper for us. I don't have time to find another tricky solution. For the moment this can't be done.

It's worth anyway since we added the multiple file selection to the current dialogs.

**Comment by Digitalone1** [comment]
I had time to work on this. Exploring the option of showing a list of community presets. I managed to make this sketch:

![image](https://github.com/wwmm/easyeffects/assets/25790525/2e426942-baed-4ae1-acfc-74f95c1e4019)

It's the input/output stack that contains two pages. Each pages (for input and output) contains another stack for local and community presets. The community stacks are empty, but the idea is to fill them with a list of installed input/output presets.

@wwmm What do you think?

**Comment by wwmm** [comment]
>What do you think?

Hum... I think that it is visually confusing to have the new tabs right below the input/output presets tabs. It is not clear at first glance what is inside of what hierarchically.

**Comment by wwmm** [comment]
>Hum... I think that it is visually confusing to have the new tabs right below the input/output presets tabs. It is not clear at first glance what is inside of what hierarchically.

Maybe if we do what someone asked at #2424 the visual confusion will go away. In this case the only tabs would be the ones related to the user/community presets.

**Comment by Digitalone1** [comment]
> Hum... I think that it is visually confusing to have the new tabs right below the input/output presets tabs. It is not clear at first glance what is inside of what hierarchically.

You're right, but I didn't know how to split the input presets from the output ones. The hovering of the top stack should help for this. Take in consideration that the community stacks will be different because they don't need the string entry for the preset name.

My idea is to have a list of community presets where each rows has an import button. The click on the button will copy the relative file in the config folder. The input presets will be copied into the input local folder, output ones into the output local folder.

**Comment by Digitalone1** [comment]
> Maybe if we do what someone asked at #2424 the visual confusion will go away. In this case the only tabs would be the ones related to the user/community presets.

If you're sure, this has to be done before the implementation of community presets.

**Comment by Digitalone1** [comment]
Or maybe a unique list of community presets and there you choose where to copy the preset? But I wonder how? Two import buttons? One for input and the other for output?

But outputs presets will always work for the input pipeline? As far as I remember, the same configuration of effects still need two different files for input and output pipelines. Am I right?

**Comment by wwmm** [comment]
>My idea is to have a list of community presets where each rows has an import button. The click on the button will copy the relative file in the config folder. The input presets will be copied into the input local folder, output ones into the output local folder.

It seems fine. The main problem right now is the list location in the window/menu.

>If you're sure, this has to be done before the implementation of community presets.

Yes. I have to find some time to take a look at the code to see how complicated or not it will be to put each section in its own menu. But now that `Baldur's Gate 3` has taken over my life I have no idea when this is happening :smile: .

>Or maybe a unique list of community presets and there you choose where to copy the preset? But I wonder how? Two import buttons? One for input and the other for output?

The list location problem remains. I think that the difficulty here is that the current menu doesn't really have the space for additional lists. Either we redesign it or put the list somewhere else.

>But outputs presets will always work for the input pipeline? 

I think that the only difference is that the preset file main section is named as `output` or `input` depending on the case. So we can't use the same file for both right now. It is this way because in the past there were really differences between the pipelines when it came to which plugins could be in each of them. But I think that nowadays there is not restriction. A plugin can be in any pipeline.

**Comment by Digitalone1** [comment]
So, what do you suggest to do now? Luckily in these days I have some free time, so I could try to implement the community presets lists. Don't know if it's the same in the next weeks. But should I try to do it in the above preset menu? Or do you suggest to wait for a new visual configuration?

I wrote an algorithm in pseudo code for the folder scanning. In the case of community presets, we can't monitor all the folders (there should be too much if the user installs many packages). So I had the idea to scan them recursively and set a limit to the subfolders. I.e. the community presets can't have more than one level of subfolders, so `output/pchan_presets/preset1.json` and `output/pchan_presets/rock/preset1.json` are allowed, but not `output/pchan_presets/rock/metal/preset1.json`.

The target is to get a map with filenames and full paths. A map is useful in this case because we want to show unique names in the list, but folders are different and filenames can be doubled, so a map ensure the uniqueness of the preset names while we track their positions (packagers should always take care of choosing a name that is not already used by other packages, otherwise some of their presets could not be shown).

At last the string indexes (the unique filename) are converted into a gtk string list for the list widget. When we pick a name inside the list, we get it's position on the file system searching in the map.

I leave it here for reference:

```c++
file ee_system_path = file("/usr/share/easyeffects");
file preset_output_path = file("/output");
file preset_input_path = file("/input");

map<string,string> output_presets_list;
map<string,string> input_presets_list;

array<string> preset_extention = ["json"];

presets_scan_directory_recursive(
    output_presets_list, ee system_path + output_presets_list, preset_extention, 2
);
presets_scan_directory_recursive(
    input_presets_list, ee system_path + input_presets_list, preset_extention, 2
);

GtkStringList output_presets_gtkstringlist = convert_map_indexes_to_gtk_string_list(output_presets_list);
GtkStringList input_presets_gtkstringlist = convert_map_indexes_to_gtk_string_list(output_presets_list);

void presets_scan_directory_recursive(map<string,string>& preset_list, 
                                      file folder,
                                      array<string> ext_array,
                                      uint sublevel) {
    for (file entry; folder.files()) {
        if (entry.is_folder() && sublevel > 0) {
            presets_scan_directory_recursive(
                preset_list, entry, ext, sublevel - 1
            );
        }
        
        if (entry.is_regular_file()) {
            string entry_ext = entry.get_extension();
            
            if (ext_array.contains(entry_ext)) {
                string filename = entry.get_filename_without_extension();
                string path = entry.get_full_path();
            
                presets_list.insert_or_assign(filename, path);
            }
        }
    }
}
```

This should work to populate the list (at the startup). Don't know how to do for updates. Since we can't use a monitor, I was thinking of putting a reload button (near the list) to reload the whole list in case a new package is installed and the user wants to see the new community presets.

Another think to point out is that this could work also for RNNoise models and Convolver impulses. Indeed I used `ext_array` on purpose for allowed extensions since for impulses it would be `["irs", "wav"]`.

At the moment I'd like to know if it's worth to try to implement those changes in the modified presets menu or it's better to wait for an alternative visual configuration.

> But I think that nowadays there is not restriction. A plugin can be in any pipeline.

Sure, but at the moment the main sections are still different for output and input, so a presets created for the output can't be used for the input, unless the user will manually change the file with a text editor. So we still need two separate sections for input and output community presets.

**Comment by p-chan5** [comment]
What do you think of a button located in the options menu that says Get Community Presets?

![003](https://github.com/wwmm/easyeffects/assets/123143444/e652d61e-c025-4258-bb51-980a4f83c331)

This button will follow the user's color scheme and redirect to https://github.com/wwmm/easyeffects/wiki/Community-Presets although if you prefer something more elaborate, a new dialog window will open listing the name of the package next to the Download button. To do this we must create a script that allows us to obtain the source code of each repository, extract it and place it in the corresponding folders.

**Comment by Digitalone1** [comment]
@p-chan5 I'd like to avoid to open Github website for this. And the script to download the source code of the repository is very hard to do (we don't even have a network stack to do download stuffs from the internet).

The problem is that we had it almost completed, but the missing feature of gtk `initial-folder` for the opening dialogs forces us to find another way to import the presets. Loading them in a specific list would be handy, but it's tricky to implement.

Making a drag and drop from the folder is even harder I think. I'm afraid there would be some issues with Wayland or Xorg. Or some Wayand implementation would work while others not.

**Comment by p-chan5** [comment]
What if we use `wget` or `curl`? [Jack's repository](https://github.com/JackHack96/EasyEffects-Presets) has an [installation script](https://github.com/JackHack96/EasyEffects-Presets/blob/master/install.sh), maybe it can help us a lot.

**Comment by p-chan5** [comment]
The dialog window would be similar to the Preferences. Only in this case the packages would be listed plus a download button (which would call these commands). The title of this window would be Download Community Presets.

**Comment by wwmm** [comment]
>At the moment I'd like to know if it's worth to try to implement those changes in the modified presets menu or it's better to wait for an alternative visual configuration.

I think it is better to make the input/output presets menu independent  of each other before trying to add one more section to the preset menu. The chances of having to work twice if we do otherwise seem high.

>Sure, but at the moment the main sections are still different for output and input, so a presets created for the output can't be used for the input, unless the user will manually change the file with a text editor. So we still need two separate sections for input and output community presets.

Under the hoods yes. But if the menu shows only output presets while the user is the the output effects tab and only input presets while the user is in the input effects tab the presets menu won't need to show a switch for input/output presets anymore. There will be space to put the community/local switch in its place.

>What if we use wget or curl? [Jack's repository](https://github.com/JackHack96/EasyEffects-Presets) has an [installation script](https://github.com/JackHack96/EasyEffects-Presets/blob/master/install.sh), maybe it can help us a lot.

This will bring with it the usual security concerns involved when calling external commands. It is better to avoid.

**Comment by wwmm** [comment]
>This will bring with it the usual security concerns involved when calling external commands. It is better to avoid.

And I am not even sure if calling an external executable would work out of the box in the PipeWire package.

**Comment by vchernin** [comment]
> And I am not even sure if calling an external executable would work out of the box in the PipeWire package.

Do you mean the flatpak package? We have a fairly standard set of executables like most distros have. However currently the flatpak package does not have any network access. We could grant it but downloading things at runtime is in my opinion best to avoid.

**Comment by wwmm** [comment]
>Do you mean the flatpak package?

Oops. Yes. That is what happens when we write answers while playing :smile:

**Comment by p-chan5** [comment]
So I've been thinking deeply about this for the last few days and I think we're making things complicated...

Your project is incredible and inspiring but unfortunately the target audience wants quick results.

As a Linux audio enthusiast user, I could join your development team. I don't know programming, but I would like to help you on the settings side. I know a bit about mixing and over time I have learned how to use the Calf and LSP plugins.

I could fork the repositories with presets, analyze them, optimize them and update them to the new version of EasyEffects. Looking at your code, it would look like [this](https://github.com/wwmm/easyeffects/tree/master/data/presets).

Please, consider implementing just a few. I'm sure end users will be happy with this decision.

**Comment by Digitalone1** [comment]
As I said before, we was near to make it, but GTK implementation of opening the dialog always at the recent files ruined our plans.

We need a way to show presets to users. Community presets can already be deployed, but the users need to know where they are stored and reach them from the opening dialog.

It's not so user friendly and my time is not so much to work on it to find a new solution. Besides there's also the implementation of new LSP features that slows down things.

We could already publish the guidelines for community presets installations where we define the locations where presets, impulses and noise models should be stored. But at the moment, since it's missing a way to show them inside the application, it's just like the already existing presets on Github. Place them wherever you want, the user should know the directory and reach them through the opening dialog since it starts always at recent files.

**Comment by p-chan5** [comment]
I really appreciate your efforts, I appreciate that you are interested in improving the app's performance, but I think this will be more confusing for novice users.

You guys have done an excellent job porting high level tools like Calf and LSP, I have no doubt about that. However, I think someone is needed who can help them distribute presets that allow the user to take full advantage of the capabilities of their hardware (speakers, monitors, headphones or microphones).

Please, consider adding them to your code. I promise to maintain them.

**Comment by p-chan5** [comment]
I would start with the [Community Presets](https://github.com/wwmm/easyeffects/wiki/Community-Presets), you can tell me which ones are your favorites and why, then I will dedicate myself to optimizing them and updating them to the latest version of Easy. Then you can try them and tell me if you need anything else. Eventually we can add them to `/usr/share/easyeffects/presets` I think 6 or 8 will be more than enough.

**Comment by wwmm** [comment]
>Please, consider adding them to your code. I promise to maintain them.

In the past I used to embed some presets in the graphical interface source code. As time passed this proved to be a bad decision. Eventually we changed plugins or the plugins evolved and having sections of the code and the presets matching was a problem. That is why those gstreamer presets we used to have are not in the window anymore. 

Another problem of having presets embedded in the app code is that they will always there there is no deleting them. So the interface will have to be redesigned exposing someway to hide all of them as it won't make sense to force them to be visible for everybody.

What would be more manageable is what I suggested some weeks ago. We could have a subfolder in our sources that through some optional git submodule magic downloads at build time the presets directly from collaborators personal repositories. Our Meson scripts could copy what is there to `/usr/share/easyeffects/presets` when `ninja install` is executed.

The advantages of this method is that I do not have to think about what is happening to the community presets. And presets authors can do their things at their own time in their personal repositories. In any case we would still have the problem described by @Digitalone1. As our window is right now there is no way for it to make the users aware of these presets existence. It would not be any different than to have package maintainers independently putting community presets inside `/usr/share/easyeffects/presets`.

**Comment by wwmm** [comment]
>I could fork the repositories with presets, analyze them, optimize them and update them to the new version of EasyEffects. Looking at your code, it would look like [this](https://github.com/wwmm/easyeffects/tree/master/data/presets).

That is more or less what I wrote in my last post. But I prefer to have them downloaded by Meson at compilation time than to have them permanently in our sources. I did not delete the old gstreamer preset files just because I completely forgot about them. They are not maintained.

And we would still have the problem about them not being visible until some redesigning is done to our window.

**Comment by Digitalone1** [comment]
If we add presets in `/usr/share/easyeffects/presets`, nothing will change from the current situation. 

* The majority of users don't know where they are located 
* if they knew it, they still have to reach them by the opening dialog starting from recent files

Besides we agreed that the community will make the packages because the developers focus on the code. If you want, you can make a package installing presets in that location.

**Comment by p-chan5** [comment]
One thing I can do is collect presets from other users, tweak them and embed them in [EasyPulse](https://github.com/p-chan5/EasyPulse), since there are many repositories hosting presets, they better all be there. This way I can keep them up to date with the latest stable version. This way I can keep them updated with the latest stable version.

It will also be easier for packagers to build EasyEffects with or without community presets.

**Comment by Digitalone1** [comment]
@p-chan5 you can already do it. One thing I suggest is to tweak them in extra versions, including also the original standalone version and mentioning the author.

Maybe in the future we could show them in the UI when the packages are installed.

At the moment we don't know how to do it. I tried weeks ago, but @wwmm suggested to wait first on syncing the preset menu on input and output effects window (for example, input presets only shows in the input tab). This should make things easier for introducing specific stack pages for community presets.

**Comment by Digitalone1** [comment]
This is going nowhere. There are other priorities now, but the biggest stopper here is where to show those community presets.

We have a schema of where to locate them (see [here](https://github.com/wwmm/easyeffects/issues/2445#issuecomment-1637137739)). We have a stub of how to retrieve and organize them in the code (see [here](https://github.com/wwmm/easyeffects/issues/2445#issuecomment-1676454034)).

The problem is that we don't have any clue where to present them to the user. Since the refactoring of the current plugin menu seems very tricky, I suggest to find another location. What about adding a button somewhere in the Effects tab and showing the list in place of the plugin UI? Every entry in the list will have an import button. After clicked this button, the preset is copied locally and it will be available in the preset menu to be applied.

**Comment by wwmm** [comment]
> What about adding a button somewhere in the Effects tab and showing the list in place of the plugin UI? Every entry in the list will have an import button.

I think I did not visualize in my mind where exactly would this be and how it would look.

>Since the refactoring of the current plugin menu seems very tricky

I am looking at this code now to see how hard this would be.

**Comment by wwmm** [comment]
>I am looking at this code now to see how hard this would be.

I have started the refactoring. Fortunately it is not as hard as I imagined. Just boring :smile:

**Comment by Digitalone1** [comment]
> I have started the refactoring. Fortunately it is not as hard as I imagined. Just boring 😄

I saw it. It's more clean. It could be easier to implement the community presets, even if there are other priorities now.

**Comment by wwmm** [comment]
>I saw it. It's more clean. It could be easier to implement the community presets, even if there are other priorities now.

Yes. It is just a matter of bringing back the tabs and making one of them showing the list of community presets. Now there is space for this.

**Comment by v-fox** [comment]
@wwmm 
> Sure. But I am not sure if these system folders are being used by many people. Even after all these years I can't remember cases where people were actually using presets in system folders.

But how would you know what people do or don't use? I was using it to package sensible defaults together with the app itself, so my secondary machines would be good to go for every user account on system-wide package installation.

Removing so-called ["deprecated"](2ea713dbc25d73143c65d6372ed7c6b7e45a1edf) functionality without any alternative to replace it with is not a good move. Distro/repo-defaults and some suggestions from community are not the same thing and they do not cancel each other. `/usr/share/<app>/presets` would be a better place for distro-packaged presets but /etc/ is better than nothing. 

Luckily, the commit still seems to be reversible, so I've removed it in my build.

**Comment by wwmm** [comment]
>But how would you know what people do or don't use? 

The only way to know is if people talk about it here. And after all these months this was the first time someone came here to talk about those system folders. What is a sign there are not many people using them.

But the amount of users isn't a problem by itself.  In order to keep things consistent in our menu we would have to complicate even more a code that is already not simple just because we can't delete files in the system path. The delete button was useless for files in the system path. The complication requried to handle this didn't feel worth the effort at the time and it still doesn't.

That being said if a better solution to handle those folders come I am not against having them back.

**Comment by Digitalone1** [comment]
Community presets will solve this issue since the file is saved in a system folder and it will be available to be imported by different users on the same device.

**Comment by Digitalone1** [comment]
Next week I should have time to work on this. I'd like to finally implement this feature we are waiting for months.

**Comment by Digitalone1** [comment]
@wwmm @vchernin Just a reminder for me. We agreed the following paths for in/output community presets:

```c++
  const std::filesystem::path input_community_preset_dir = "/usr/share/easyeffects/input";
  const std::filesystem::path output_community_preset_dir = "/usr/share/easyeffects/output";
```

So, on a regular installation (i.e. upstream Arch build), the app will look for packages installed under `/usr/share/easyeffects/...`.

But for Flatpak builds, the app will look for custom directories under `/var/lib/flatpak/...`, even if the source code contains `/usr/share/easyeffects/`, right?

**Comment by Digitalone1** [comment]
@wwmm I like this widget: https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1.5/class.ActionRow.html

My idea is to emphasize the community preset name, still showing it's related package as subtitle. But I'm afraid this widget can't be used inside a `GtkListView`. Do you think it's possible?

To make things easier for me, I'd like to copy the existing local presets structure for the community presets. But I'm not sure if other widgets can be used rather than the GtkBox inside `preset_row.ui`.

Or maybe it can be used inside a GtkBox, but since it's not used inside a `GtkListBoxRow`, all the borders will be rounded?

**Comment by wwmm** [comment]
> @wwmm I like this widget: https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1.5/class.ActionRow.html
> 
> My idea is to emphasize the community preset name, still showing it's related package as subtitle. But I'm afraid this widget can't be used inside a `GtkListView`. Do you think it's possible?
> 
> To make things easier for me, I'd like to copy the existing local presets structure for the community presets. But I'm not sure if other widgets can be used rather than the GtkBox inside `preset_row.ui`.
> 
> Or maybe it can be used inside a GtkBox, but since it's not used inside a `GtkListBoxRow`, all the borders will be rounded (instead of only the first and the last row as in the `AdwPreferencesGroup`)?

I think that it is possible to put it inside a GtkListView. But I am not sure if it will look the same way it usually does when put inside libadwaita containers like the ones used in the other places where we used a AdwActionRow. 

>Update: Besides, do you think it's better to use GtkStack or [AdwViewStack](https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1-latest/class.ViewStack.html)?

A long time ago I saw some differences when trying to interact with the stack from the C code side. But I do not remember what it was now. As the preset menu is already using an AdwViewStack try to use it too. If at some point we see a disadvantage we can switch to the other.

**Comment by wwmm** [comment]
> > As the preset menu is already using an AdwViewStack try to use it too.
> 
> Are you sure? I don't see an `AdwViewStack` inside the `presets_menu.ui`. Maybe it was present in the previous releases when the presets menu had the stack switcher between input and output.

After a quick search in vscode I've found https://github.com/wwmm/easyeffects/blob/281a508970fa34829239d0a9127cae71d9e1f58c/src/presets_menu.cpp#L58. But now take I am looking at the ui file I do not see a stack there. IT is probably an useless variable that was forgotten to be deleted.

**Comment by wwmm** [comment]
But back to the main topic. It is possible that there are style differences between `AdwViewStack` and `GtkStack`. It will probaly be necessary to test both to see what fits better.

**Comment by Digitalone1** [comment]
Yes, I confirm it's an unused variable. I already removed it. Anyway the main difference is written [here](https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1-latest/class.ViewStack.html): Unlike [GtkStack](https://docs.gtk.org/gtk4/class.Stack.html), transitions between views are not animated.

**Comment by vchernin** [comment]
> Just a reminder for me. We agreed the following paths for in/output community presets:
> 
> ```c++
>   const std::filesystem::path input_community_preset_dir = "/usr/share/easyeffects/input";
>   const std::filesystem::path output_community_preset_dir = "/usr/share/easyeffects/output";
> ```
> 
> So, on a regular installation (i.e. upstream Arch build), the app will look for packages installed under `/usr/share/easyeffects/...`.
> 
> But for Flatpak builds, the app will look for custom directories under `/var/lib/flatpak/...`, even if the source code contains `/usr/share/easyeffects/`, right?

Actually looking a bit more I think we want to use the `XDG_DATA_DIRS` env variable in regular installations and flatpak builds. It seems wrong to hardcode `/usr` since some distros may not use it. As far as I understand this the "standard" way of doing things.

In addition to `XDG_DATA_DIRS`, we should also search for `/app/extensions/easyeffects` when in a flatpak. We can check if in a flatpak either by checking for the existence of `/.flatpak-info` at runtime or via a build option. There is no problem with also checking `XDG_DATA_DIRS` in flatpak even if it is unlikely presets are there.

For example on my Fedora 39 system:
```bash
echo $XDG_DATA_DIRS
/var/home/vchernin/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share:/usr/local/share/:/usr/share/
```
Then each directory should be checked for `easyeffects/input` and `easyeffects/output`, e.g. `/usr/share/easyeffects/input`.  

Making it a list of directories is more complicated than one though, so I'm not sure if it is worth the trouble. But it might be worth doing the "standard" thing here like how the config directory is controllable thanks to `XDG_CONFIG_HOME`.

**Comment by wwmm** [comment]
>Making it a list of directories is more complicated than one though, so I'm not sure if it is worth the trouble. 

The problem is what to do when a preset with the same name is available in more than one path.  We discussed this in the past and I think we never got to a decent solution. But it makes sense to somehow use the XDG directories.

**Comment by vchernin** [comment]
Apparently we can just pick whichever was first in the order. There is also `XDG_DATA_HOME` which should be treated as the first/preferred directory.

> The order of base directories denotes their importance; the first directory listed is the most important. When the same information is defined in multiple places the information defined relative to the more important base directory takes precedent. The base directory defined by $XDG_DATA_HOME is considered more important than any of the base directories defined by $XDG_DATA_DIRS. The base directory defined by $XDG_CONFIG_HOME is considered more important than any of the base directories defined by $XDG_CONFIG_DIRS. 

https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html (version 0.8)

**Comment by Digitalone1** [comment]
> Actually looking a bit more I think we want to use the `XDG_DATA_DIRS` env variable in regular installations and flatpak builds. It seems wrong to hardcode `/usr` since some distros may not use it. As far as I understand using `XDG_DATA_DIRS` is the "standard" way of doing things.

That can be done. But two questions come to my mind.

1. How can I get those directory list in C++?
2. We have to write guidelines for packagers telling them where to install these presets. So I think we will end up to search only in one location. The point is: which directory? If you will make a Flatpak package with EE presets, where will you install them?

> Apparently we can just pick whichever was first in the order. There is also `XDG_DATA_HOME` which should be treated as the first/preferred directory.

We already use this for the local presets. And there's a Gtk function to get this directory. But I don't think the presets will be installed locally (under `/home/user/.local/...`), at least not in regular distributions.


> The problem is what to do when a preset with the same name is available in more than one path. We discussed this in the past and I think we never got to a decent solution. But it makes sense to somehow use the XDG directories.

Well, if we make the ListView using the full path, there would not be an issue.

**Comment by vchernin** [comment]
> 1. How can I get those directory list in C++?

There is a glib function https://docs.gtk.org/glib/func.get_system_data_dirs.html. Not C++ but probably still sensible given it handles fallback like the spec says.


> 2. We have to write guidelines for packagers telling them where to install these presets. So I think we will end up to search only in one location. The point is: which directory? If you will make a Flatpak package with EE presets, where will you install them?

I think using `XDG_DATA_DIRS` will work fairly elegantly, as if a distro is packaging application in data in `/foo` `XDG_DATA_DIRS` must also contain `/foo`, it would be very strange otherwise and sounds like a distribution bug.

So besides "package the presets in the standard application data directory" I don't think it is possible to give more specific instructions for distributions. 

For arch is is fairly clear `/usr/share/pkg`  i.e. `/usr/share/easyeffects` should be used for "application data". https://wiki.archlinux.org/title/Arch_package_guidelines#Directories. Other distros probably also document these things or at least adhere to a sensible ish default. 

How to package flatpak presets will definitely need their own example and documentation. In practice the only directory flatpak apps will care about is `/app/extensions/easyeffects`, so that one can be documented but it is not wrong to check `XDG_DATA_DIRS` as well.

> We already use this for the local presets. And there's a Gtk function to get this directory. But I don't think the presets will be installed locally (under `/home/user/.local/...`), at least not in regular distributions.

Actually I misread the documentation. `XDG_DATA_HOME` is only meant for user-set data as indeed EE is already using it. Packages should only install to somewhere in `XDG_DATA_DIRS` which is meant only for system-wide data (which is what the linked glib function does).

**Comment by Digitalone1** [comment]
@vchernin Okay.

This is a pita because some paths end with `/`, others without, it's not easy to append the `input`/`output` folder, but I will come up with a solution.

**Comment by Digitalone1** [comment]
@wwmm  A first incomplete draft:

![image](https://github.com/wwmm/easyeffects/assets/25790525/e9d634b3-038a-48d1-88cf-39175b995c98)

**Comment by wwmm** [comment]
> @wwmm A first incomplete draft:
> 
> ![image](https://private-user-images.githubusercontent.com/25790525/317507134-e9d634b3-038a-48d1-88cf-39175b995c98.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1ODIwMzMsIm5iZiI6MTcxMTU4MTczMywicGF0aCI6Ii8yNTc5MDUyNS8zMTc1MDcxMzQtZTlkNjM0YjMtMDM4YS00OGQxLTg4Y2YtMzkxNzViOTk1Yzk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI3VDIzMjIxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRkNjM0MmIyMTU4ZTNmMjYxMzNkMmVhMTQ3YWIwOTRmODdiOWZkNmRmODc3ZjU5ZjgwYjQyYzNkNWQ2M2E3MzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GEnKStHH6NQVRNDhb7JpEQ1XEi5kRl8a1C2nWRpfmXg)

It seems good to me.

**Comment by Digitalone1** [comment]
@wwmm The list is indexed by the full paths. This means the same preset name is listed twice if it's contained in different packages. In example:

- /xdg_data_dir/easyeffects/output/my_package/rock.json
- /xdg_data_dir/easyeffects/output/another_package/rock.json

Every item in the list reports the package name in the top label, so the user can still distinguish them.

But the problem with this solution is that the search does not work as intended in local presets list. Because the full path is indexed, so it looks inside the full string. I.e `output` returns all the items because the substring is contained in the full path.

If we want to use the search bar inside the community stackpage, is there a way to look only inside the latest part of the `string-object`? Maybe a callback that can be used for the gtk widget that returns only a substring retrieved with regex. I tried to look about this, but I didn't find anything. Maybe you know better then me...

**Comment by Digitalone1** [comment]
@wwmm No, unfortunately it's not enough. We should search only in the filename part of the string-object item. In other words, applying this regex: https://regex101.com/r/frOb6Q/1

**Comment by wwmm** [comment]
> @wwmm No, unfortunately it's not enough. We should search only in the filename part of the string-object item. In other words, applying this regex: https://regex101.com/r/frOb6Q/1

Then the only way will be creating a custom model instead of using a simple `GtkStringList` as model. In other words something similar to what is done inside `apps_box.cpp` and its ui. A node holder similar to https://github.com/wwmm/easyeffects/blob/master/src/node_info_holder.cpp will be necessary. It would need one property for the full path and another only for the file name. Then in the xml we would use this holder object and property in https://github.com/wwmm/easyeffects/blob/812db4f15d9f316565cd3cbadff954c6404ae0b0/data/ui/presets_menu.ui#L105.

It is going to complicate things but I do not have a better idea at the moment.

**Comment by wwmm** [comment]
Unless there is a way to create a GtkExpression based on regex. If it is possible we could try to somehow chain one expression on another.

**Comment by wwmm** [comment]
In any case proceed as things are right now. Once I have the merged code on my computer I could try to see if there is something less annoying than creating a custom model.

**Comment by wwmm** [comment]
I am looking at some gtk4 test code and I think there is a simpler way to handle this problem https://github.com/pop-os/gtk4/blob/f2b105bd9d64f8b47b5087885907311efdf2a49b/testsuite/gtk/filter.c#L264. It is possible to create an expression for the filter from a closure. It should be possible to apply a regex filter inside of it to get only the file name.

So instead of setting `GtkFilterListModel`'s filters property in the xml as we do right now we could set it through C code and define the closure there.

**Comment by Digitalone1** [comment]
@wwmm Nice, this should be the way. I will try it out.

**Comment by Digitalone1** [comment]
At the moment I implemented the loading of a community preset to test it without importing it locally. It's Just for testing how it sounds. Then I will implement the import function.

I was thinking of avoiding to overwrite the local presets, so maybe it's better to rename the preset if there's already an existing one. In example, if `rock.json` already exists, we will rename it `rock-1.json`.

Then I will try the expression filter.

The community presets needs to be tested for a long time, so I suggest to do one of the following:
1. prior to the PR, you make a specific branch to merge everything there;
2. make a new release, so the PR can be merged into the master branch.

**Comment by Digitalone1** [comment]
@wwmm Looking at the code you linked, the following is the callback:

```c++
static char *
get_string (gpointer object)
{
  return g_strdup_printf ("%u", GPOINTER_TO_UINT (g_object_get_qdata (object, number_quark)));
}
```

But what is the `object` passed? Could I do the following?

```c++
auto* string_object = GTK_STRING_OBJECT(g_object_get_data(G_OBJECT(object), "string-object");

auto* preset_path = gtk_string_object_get_string(string_object);

```

**Comment by wwmm** [comment]
>But what is the object passed? Could I do the following?

I think it should work. That is what the gtkexpression code in our ui file is doing.

**Comment by Digitalone1** [comment]
Meanwhile I discovered a funny thing. The GTK bug that forced us to make this new system for community presets is resolved.

So now the "open file dialog" is opening in the directory passed as argument:

```c++
auto* init_folder = g_file_new_for_path(SYSTEM_PRESETS_DIR);
``` 

And looking for `SYSTEM_PRESETS_DIR`, I get the following in the `meson.build`

```
system_presets_dir = join_paths(datadir, meson.project_name() + '/presets')
``` 

Which is `/usr/share/easyeffects/presets`, the directory where we wanted to put the community presets to be imported locally.

But, anyway, since we deprecated this solution in favor of the new plugin menu UI, I assume `SYSTEM_PRESETS_DIR`, `SYSTEM_IRS_DIR`, and `SYSTEM_RNNOISE_DIR` can be removed from `meson.build`. And we can open the file dialog in another folder (the user folder I guess?).

**Comment by wwmm** [comment]
>Meanwhile I discovered a funny thing. The GTK bug that forced us to make this new system for community presets is resolved.

It happens :smile: 

>But, anyway, since we deprecated this solution in favor of the new plugin menu UI, I assume SYSTEM_PRESETS_DIR, SYSTEM_IRS_DIR, and SYSTEM_RNNOISE_DIR can be removed from meson.build. And we can open the file dialog in another folder (the user folder I guess?).

I think we can remove them. I do not remember anything else needing these variables.

**Comment by Digitalone1** [comment]
> I think we can remove them. I do not remember anything else needing these variables.

Ok, I managed to remove them.

**Comment by Digitalone1** [comment]
No luck. I tried to set the new expression in different ways, but the search does not work. Or rather, it works wrong. When I start typing, all the items disappear.

Inserting a debug line inside the callback it's showing nothing on the command line, it's not even called.

There's nothing I can do on this side. I miss only the implementation of the import button callback.

But tomorrow I could make a PR in case you want to test it and fix the string filter.

@wwmm What do you want to do? A new branch or a new release to merge into the master?

**Comment by wwmm** [comment]
>What do you want to do? A new branch or a new release to merge into the master?

As the new code is just adding a new tab to the new menu it should not be problematic to do a PR directly to our master branch. It will be easier to have more people testing it this way than keeping the whole thing in a different branch most people won't know it exist.

>No luck. I tried to set the new expression in different ways, but the search does not work. Or rather, it works wrong. When I start typing, all the items disappear.

I will try to do some tests after we merge the PR.

**Comment by Digitalone1** [comment]
> As the new code is just adding a new tab to the new menu it should not be problematic to do a PR directly to our master branch. It will be easier to have more people testing it this way than keeping the whole thing in a different branch most people won't know it exist.

Ok, then make a new `7.1.6` release which fixes the level bars. I suggest the next one with community presets to be `7.2.0`.

**Comment by Digitalone1** [comment]
Implemented the import function. I miss only the refresh button to show new presets in case a new package is installed. After this, I will make the PR.

**Comment by Digitalone1** [comment]
@wwmm PR made.

What should be completed:

- [ ] The method to filter the community presets by the search entry (a GtkExpression closure is used, but it's not working; I don't even know if it has to be freed after binding it, please check it out)
- [ ] The visibility of the StatusPage when the community presets list is empty. I'd like to show a message to the user, but I didn't have time to complete it (and honestly I don't even remember how it should be done, I only added it into the builder).
- [ ] A guideline for packagers should be written suggesting some rules to follow for the correct installation of the packages.

@vchernin @p-chan5 @v-fox @servimo please test and, if you can, make a test package on Flatpak.

**Comment by wwmm** [comment]
>The method to filter the community presets by the search entry (a GtkExpression closure is used, but it's not working; I don't even know if it has to be freed after binding it, please check it out)

It is almost like gtk is ignoring the call. Really strange...

**Comment by servimo** [comment]
@Digitalone1 I don't know what to do, I am not a dev. I compiled it and Community Presets is there, a blank page. 
I don't use Easyeffects flatpak.
 I was with "Dengue". Sorry for this.

**Comment by Digitalone1** [comment]
> @Digitalone1 I don't know what to do, I am not a dev. I compiled it and Community Presets is there, a blank page. I don't use Easyeffects flatpak.



> @Digitalone1 I don't know what to do, I am not a dev. I compiled it and Community Presets is there, a blank page. I don't use Easyeffects flatpak.

@servimo Copy some presets in the following paths:
- xdg_data_dir/easyeffects/output/my_package/rock.json
- xdg_data_dir/easyeffects/output/another_package/rock.json

`echo $XDG_DATA_DIRS` from command line should list the paths where xdg_data_dir are. There should be one easily accessible under ~/.local//share/flatpak/...

Then click the Reshesh button or just restart the application.

**Comment by servimo** [comment]
ok, I copy a preset to /usr/share/easyeffects/Teste/output/ and now it appears in the Community Presets (preset name Try / Import)

**Comment by Digitalone1** [comment]
> My question is how to bundle with the .irs file?

At the moment only plugin presets are supported. Maybe in the future also irs and rnnoise models will be supported.

**Comment by wwmm** [comment]
I've found a way to make the community preset search to work.

**Comment by servimo** [comment]
What about if can add xdg_data_dir/.local/share/easyeffects/output/my_package/rock.json?
So I can create a favorite folder in there and use it as some favorite presets?

**Comment by Digitalone1** [comment]
> What about if can add /home/user/.local/share/easyeffects/output/? So I can create a favorite folder in there and use it to store some favorite presets?

You can use `/home/user/.config/easyeffects/output`.

**Comment by servimo** [comment]
Am I being annoying? I already ask for this and I was thinking I can load a preset when I insert a certain device in the USB, so yes I am being annoying. You don't need to do this, the utility to Community Presets is another.

**Comment by Digitalone1** [comment]
Since the search has been fixed, we miss only a document with guidelines for packagers.

**Comment by vchernin** [comment]
> At the moment only plugin presets are supported. Maybe in the future also irs and rnnoise models will be supported.

I've been looking at making an example flatpak extension along with some documentation. One thing to consider is the name of the "extension point" which is basically an ID the easyeffects flatpak and all the flatpak extensions need to agree on in order for preset files to be mounted in the easyeffects flatpak filesystem. Right now it is `io.github.wwmm.easyeffects.Presets`. But would such a name make sense for rnnoise models or irs files? 

I don't know if there is a good word to describe everything we'd want to support, besides perhaps `Plugins` i.e. `io.github.wwmm.easyeffects.Plugins`. It would be a bit strange to have multiple extension points like having `io.github.wwmm.easyeffects.Presets` and `io.github.wwmm.easyeffects.Models`, as one flatpak extension package can only support one at a time. 

So would it be sensible for the flatpak system to refer to these packages as "plugins" for future proofing?

Another thing the flatpak system lets us do is versioning. It could be possible to have an extension that publishes extension versions for both easyeffects 7.x.x. and 8.x.x. But I don't see a huge reason for that as for the most part easyeffects should remain compatible, and because these are (for now) just json files a versioning scheme would be more confusing than helpful. Things like OBS studio plugins on flathub don't maintain different versions and OBS plugins are considerably more complex. So I can make it more or less "versionless".

**Comment by wwmm** [comment]
>So would it be sensible for the flatpak system to refer to these packages as "plugins" for future proofing?

I think that "plugins" will give people the wrong idea. The word "presets" is not a good match but it is better than plugins in this case.

Maybe we could use "extra" or "community" there.

**Comment by Digitalone1** [comment]
I'm not very good in English, but I think `presets` could be good since models and irs can be considered as "presets" for RNNoise and Convolver respectively.

**Comment by Digitalone1** [comment]
@vchernin as I pointed out in #3041, the guidelines needs also instructions for the flatpak ID that the extensions should have in order for preset files to be mounted in the easyeffects flatpak filesystem.

Feel free to add those details here: https://github.com/wwmm/easyeffects/blob/master/COMMUNITY_PRESETS_GUIDELINES.md

For the moment we have the guidelines in the repo. Then we will see if it's worth moving it in the wiki.

**Comment by Digitalone1** [comment]
There's a problem with Convolver irs and RNNoise models.

The presets save the path where to load them but, as it is now, this mechanism gives various issues.

1. If the preset is exported on another system, you should have the same path, in other words the same username, otherwise the file is not found.
2. Given that mentioned above, using a community preset with a Convolver or a Noise Reduction effect is useless, the files are not loaded in both cases of Trying and Importing.
3. This design of saving the path it's not the best because if I want to share my preset, other users will know the username on my system.

@wwmm What can we do to make this work properly? I was thinking of saving only the filename and lookup it in local and xdg_data folders. But we should also found a method to import the files when clicking the import button inside the preset menu popover.

**Comment by wwmm** [comment]
>What can we do to make this work properly? I was thinking of saving only the filename and lookup it in local and xdg_data folders. But we should also found a method to import the files when clicking the import button inside the preset menu popover.

I think this is the only way to improve this problem we have since forever. It should not cause much trouble to start the search in the local path and falling back to other xdg_data paths if the file is not found. 

What exactly is the problem about the file importing? Is there the need to change this?

**Comment by Digitalone1** [comment]
> What exactly is the problem about the file importing? Is there the need to change this?

If the user try or import the community preset with a convolver/rnnoise, it should work because the filename reported into the preset will be found (it's installed in the package path... it's responsibility of the packager installing it in the right place, following our guidelines...).

The problem is when the user imports the preset and then uninstall the package. The local preset won't work anymore because the file does not exist. We should import the irs/model locally alongside the preset. Can we look inside the preset from PresetsManager to search for irs/model filenames and copy them?

There's also the other issue that if we start to use the filename rather than the path, the existing presets are deprecated (and not working anymore, except if we make some workaround...)

**Comment by wwmm** [comment]
>Can we look inside the preset from PresetsManager to search for irs/model filenames and copy them?

Hum... It is probably better to make the community preset import routine to also copy the impulse/rnnoise files to the local folder if it detects the preset contains a reference to these files. THis way there won't be the need to change things in presets_manager.

>There's also the other issue that if we start to use the filename rather than the path, the existing presets are deprecated (and not working anymore, except if we make some workaround...)

Hum... At this moment the key in the presets is named `kernel-path`. Maybe we could add a new key like `kernel-file-name` and still try to read the path key for backwards compatibility. Newer presets would only set the one about the file name. But we could still check if the one about the path is there to still load old presets.

**Comment by Digitalone1** [comment]
> Hum... It is probably better to make the community preset import routine to also copy the impulse/rnnoise files to the local folder if it detects the preset contains a reference to these files. THis way there won't be the need to change things in presets_manager.

And how can we detect if the preset contains references to this files?

**Comment by Digitalone1** [comment]
Never mind. I started to make a refactoring and it seems doable. I think it's best to keep both `kernel-path` and `kernel-name` since various parts of the code rely on `kernel-path`. We can just save only the `kernel-name` inside the preset and keep `kernel-path` in gsettings since we derive it starting from the `kernel-name` and use it in various part of the code rather than constructing it every time...

**Comment by wwmm** [comment]
>Maybe in the preset reading stage extracting the name from the path and saving to kernel-name. @wwmm what do you think?

I think it is fine. In the past something similar happened when we had to change the format of the preset file. It should be enough to do the parsing only in the reading phase. When writing the new format will be used anyway.

**Comment by Digitalone1** [comment]
Waiting for a real package to try, especially on Flathub, I'd like to point out that it's better to test it before making a new release.

**Comment by vchernin** [comment]
I've pushed backwards compatible packages changes to support the flatpak presets on flathub https://github.com/flathub/com.github.wwmm.easyeffects/commit/6e302cdc5d8ad9314fa7733df48f67749551a491.

@Digitalone1 I don't mind sending a pr to flathub to create a package for your preset repo to do the first one. I don't want it to be added to flathub just yet, but it would be good to do it sooner rather than later.

**Comment by wwmm** [comment]
>How can I test the master branch? 

I think you can download one of the artifacts generated by our github [actions](https://github.com/wwmm/easyeffects/actions) and install it manually through the command line. I think an user did this in the past.

**Comment by vchernin** [comment]
> Looking the presets listed in the Wiki, I suggest to mix my preset with [JackHack's](https://github.com/JackHack96/EasyEffects-Presets/), since one of them rely on a [irs file](https://github.com/JackHack96/EasyEffects-Presets/blob/master/Bass%20Enhancing%20%2B%20Perfect%20EQ.json#L7) which gives us the opportunity to test the backward compatibility to the previous presets.

I considered making a preset package mixing both preset repos, together, but I decided it was strange and just made 2 
preset packages, one for your LoudnessEqualizer and one for all JackHack96's presets. 

To test it locally indeed you can first install the devel build from github actions. https://github.com/wwmm/easyeffects/actions/runs/8786350683/artifacts/1436075826. Then you can build the preset packages using `flatpak-builder` with the flathub repo branches I have:

```bash
git clone --branch=io.github.wwmm.easyeffects.Presets.JackHack96 https://github.com/vchernin/flathub flathub-JackHack96
flatpak-builder build-dir-JackHack96 --user --install flathub-JackHack96/io.github.wwmm.easyeffects.Presets.JackHack96.json --force-clean

git clone --branch=io.github.wwmm.easyeffects.Presets.LoudnessEqualizer https://github.com/vchernin/flathub flathub-LoudnessEqualizer
flatpak-builder build-dir-LoudnessEqualizer --user --install flathub-LoudnessEqualizer/io.github.wwmm.easyeffects.Presets.LoudnessEqualizer.json --force-clean
```

**Comment by vchernin** [comment]
@Digitalone1 how are IRS files supposed to work within the UI? Output and input presets are clearly working on my machine, I can see them in the top left menu. But I cannot figure out where the IRS should show up in the convolver. The IRS file is definitely present at the correct location.

If it is supposed to show up in this menu it is missing.
![image](https://github.com/wwmm/easyeffects/assets/38842733/8c69a2a0-9174-4254-86dd-10fb306b2c6f)

**Comment by vchernin** [comment]
> how are IRS files supposed to work within the UI? Output and input presets are clearly working on my machine, I can see them in the top left menu. But I cannot figure out where the IRS should show up in the convolver. The IRS file is definitely present at the correct location.
> 
> If it is supposed to show up in this menu it is missing.

Oh, now I realize the IRS files are installed with the preset, nevermind.

**Comment by vchernin** [comment]
Hm, following the instructions above when trying to import JackHack96 `Bass Boosted.json` I see:

```
(easyeffects:2): easyeffects-WARNING **: 15:42:50.414: 	presets_manager.cpp:850	[json.exception.out_of_range.403] key 'kernel-name' not found

(easyeffects:2): easyeffects-WARNING **: 15:42:50.414: 	presets_manager.cpp:922	can't import addons for the community preset: /app/extensions/Presets/output/JackHack96/Bass Boosted.json; import stage aborted, please reload the community preset list

(easyeffects:2): easyeffects-WARNING **: 15:42:50.414: 	presets_manager.cpp:925	if the issue goes on, contact the maintainer of the community package
```

Which sounds like because this is an older preset? https://github.com/JackHack96/EasyEffects-Presets/blob/8a4499d34e6c461ecda548f077d49ddf8ba73e12/Bass%20Boosted.json#L17

**Comment by Digitalone1** [comment]
@vchernin It seems `/app/extensions/Presets` path should be added also in other parts of the code. Try to look inside the Convolver and the RNNoise classes.

**Comment by Digitalone1** [comment]
@vchernin I managed to install the artifact, but I can't for the extension:

```
error: org.freedesktop.Sdk/x86_64/23.08 not installed
Failed to init: Unable to find sdk org.freedesktop.Sdk//23.08 version stable
```

It does not seem really straightforward. Please, test it yourself.

**Comment by vchernin** [comment]
It seems as of 35b69a75ffd72d55455dc33c3cb7b7215c040945 the "try community preset" button works (which I did not test before), but the import button still does not work with the same error.

> I managed to install the artifact, but I can't for the extension:

This error was because a dependency was missing, I modified the `flatpak-builder` [command above](https://github.com/wwmm/easyeffects/issues/2445#issuecomment-2070585018) so now it should auto-install dependencies.

**Comment by Digitalone1** [comment]
@vchernin Can't test it now, but it's weird the try function is working while the import not. Are you sure the irs file is loaded with try?

**Comment by Digitalone1** [comment]
@vchernin Maybe I understood which is the issue. The backward compatibility for `kernel-path` has been only added in the local preset loading, not in the community importing stage and, as I think about it, it's better we don't add it there since packagers should follow the latest specs.

Please, test with the updated preset. When you import the preset with the Convolver, you should have it in the local folder with a new format having the updated `kernel-name` which reports only the file name, not the full path.

Install this preset (maybe updating the repository), and it should work.

**Comment by vchernin** [comment]
> @vchernin Maybe I understood which is the issue. The backward compatibility for `kernel-path` has been only added in the local preset loading, not in the community importing stage and, as I think about it, it's better we don't add it there since packagers should follow the latest specs.

That seems reasonable enough. Although would we want to have both `kernel-name` and `kernel-path` in preset repos to have backwards compatibility for older easyeffects versions?

In the meantime for testing I pushed bass boosted with `kernel-name` to my flathub repo, from the instructions above.

> 
> Please, test with the updated preset. When you import the preset with the Convolver, **just save it** and you should have it in the local folder with a new format having the updated `kernel-name` which reports only the file name, not the full path.
> 

Indeed trying a preset and then saving it to a file works and produces kernel-name correctly. 

I also found the irs directory was not always present which is fixed by https://github.com/wwmm/easyeffects/pull/3103.

With both these fixes I can import, and then load a preset with an irs file in flatpak without any issues.

**Comment by Digitalone1** [comment]
> That seems reasonable enough. Although would we want to have both `kernel-name` and `kernel-path` in preset repos to have backwards compatibility for older easyeffects versions?

It's not fully possible. We can't report the full path in the preset. It's also an issue about privacy. The packager can still have additional presets for older versions, that are not installed in the Flatpak extension.

**Comment by Digitalone1** [comment]
@vchernin I wonder what happens if I install a Flatpak extension and use an EasyEffects regular build made by a distro. I shouldn't see those community presets, right? Presets in Flatpak extensions would be visible only in the Flatpak version. Where exactly are they stored?

**Comment by vchernin** [comment]
> I wonder what happens if I install a Flatpak extension and use an EasyEffects regular build made by a distro. I shouldn't see those community presets, right? Presets in Flatpak extensions would be visible only in the Flatpak version. Where exactly are they stored?

Indeed you will not see those community presets. Flatpak stores all flatpak package files in its own directory on the host system, unless the regular EasyEffects is aware of where to look Flatpak packages will not have any effect. Flatpak will not put anything very interesting on the host `XDG_DATA_DIRS`.

For Flatpaks installed for a singular user they will be installed to `~/.local/share/flatpak`. For a system wide install it is `/var/lib/flatpak`.

So e.g. for a user installation `~/.local/share/flatpak/runtime/io.github.wwmm.easyeffects.Presets.JackHack96/x86_64/stable/active/files` shows the files packaged in the extension.

**Comment by vchernin** [comment]
I have not seen any new issues as of https://github.com/wwmm/easyeffects/pull/3111. @Digitalone1 if you think this is the right time I can submit your preset package to flathub. 

There doesn't need to be a Easy Effects release yet, it would be good to make sure flathub doesn't have any issues with e.g. the naming convention before anything is released.

**Comment by Digitalone1** [comment]
Yes @vchernin, it would be very appreciated.

**Comment by Digitalone1** [comment]
So it seems Flathub won't likely accept presets packages. Luckily distro packages exist and they could be made in the future.

If this is the case, I will remove any mention to Flatpak/Flathub. Maybe an alternative could be creating a custom Flatpak remote, but I don't have the will, nor the knowledge to do it.

**Comment by Digitalone1** [comment]
@wwmm It's almost two months community presets have been implemented. Is there something else stopping a new release? Many things changed in the last months, it would be a good opportunity to push them to regular users.

**Comment by wwmm** [comment]
> @wwmm It's almost two months community presets have been implemented. Is there something else stopping a new release? Many things changed in the last months, it would be a good opportunity to push them to regular users.

No special reason. But I think it would be good to try to fix #3188 before doing it. I will test today's commit for a few days and make a new release by the end of this week.

**Comment by Digitalone1** [comment]
I think this can be finally closed.

@wwmm you can unpin it.
